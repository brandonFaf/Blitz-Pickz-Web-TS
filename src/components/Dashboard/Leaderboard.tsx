import React, { useEffect } from 'react';
import styled from 'styled-components/macro';
import ProfilePhoto from '../Profile/ProfilePhoto';
import Colors from '../../Styles/colors';
import useUser from '../../hooks/useUser';
import useGroup from '../../hooks/useGroup';
import { useLeaderboardLazyQuery } from '../../types/graphql.types';

const Row = styled.div`
  display: grid;
  grid-template-columns: 5vw 45px 28vw 10vw;
  grid-auto-flow: row;
  grid-column-gap: 12px;
  align-items: center;
  .current {
    color: ${Colors.highlight};
  }
`;

const LBoard = styled.div`
  display: grid;
  grid-auto-flow: row;
  margin-top: 40px;
  font-size: 14px;
  justify-content: center;
  text-align: center;
  align-items: center;
  grid-row-gap: 10px;
  overflow-y: scroll;
  margin: 0 10vw;
  grid-auto-rows: max-content;
`;

const Leaderboard = () => {
  const { user } = useUser();
  const { group } = useGroup();
  const [
    loadLeaderboard,
    { data, loading, error, called }
  ] = useLeaderboardLazyQuery();
  useEffect(() => {
    if (group) {
      loadLeaderboard({
        variables: { group_id: group.id }
      });
    }
  }, [group, loadLeaderboard]);
  if (!called || loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    console.log(error);
    return <div>There was a problem loading this data</div>;
  }
  if (!data || !data.group_by_pk) {
    return <div></div>;
  }
  const members =
    data.rankings.length === 0
      ? group?.members.map(m => ({ ...m, rank: 1, points: 0 }))
      : data.rankings;
  return (
    <LBoard data-testid='leaderboard'>
      <Row>
        <div />
        <div />
        <div>Player</div>
        <div>Points</div>
      </Row>
      {members?.map(m => {
        if (!m.user) {
          return <></>;
        }
        const { photo_url, display_name, id } = m.user;
        let cn = '';
        if (id === user.id) {
          cn = 'current';
        }
        return (
          <Row data-testid='leaderboard-row' key={id} className={cn}>
            <div>{m.rank || 1}.</div>
            <div>
              <ProfilePhoto
                size={'small'}
                displayName={display_name}
                src={photo_url}
              />
            </div>
            <div className={cn}>{display_name}</div>
            <div>{m.points || 0}</div>
          </Row>
        );
      })}
    </LBoard>
  );
};

export default Leaderboard;
