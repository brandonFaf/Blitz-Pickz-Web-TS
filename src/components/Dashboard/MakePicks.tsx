import React, { useEffect } from 'react';
import styled from 'styled-components/macro';
import ActionButton from '../../Styles/Shared/ActionButton';
import { Link } from 'react-router-dom';
import chevron from '../../img/Chevron.png';
import { usePicksLazyQuery } from '../../types/graphql.types';
import getCurrentWeek from '../../helpers/currentWeek';
import useUser from '../../hooks/useUser';
import useGroup from '../../hooks/useGroup';
const Container = styled.div`
  width: 100%;
  display: flex;
  justify-content: center;
`;

const MakePicks = () => {
  const { user } = useUser();
  const { group } = useGroup();
  const week = getCurrentWeek();
  const [loadPicks, { data, loading, called }] = usePicksLazyQuery();
  useEffect(() => {
    if (group) {
      console.log({ user: user.id, group: group.id ?? 0, week });
      loadPicks({
        variables: { user_id: user.id, group_id: group?.id ?? 0, week }
      });
    }
  }, [group, loadPicks, user.id, week]);
  if (!called || loading) {
    return <div>Loading</div>;
  }
  console.log('this is the data:', data);
  const picks = data?.user_by_pk?.picks_aggregate.aggregate?.count ?? 0;
  const totalGames = data?.game_aggregate.aggregate?.count ?? 0;
  return (
    <Container>
      <Link data-testid={'make-picks'} to='/picks'>
        <ActionButton>
          <img src={chevron} alt='chevron' />
          <div>MAKE YOUR PICKS</div>
          <div>{`${picks}/${totalGames}`}</div>
        </ActionButton>
      </Link>
    </Container>
  );
};

export default React.memo(MakePicks);
