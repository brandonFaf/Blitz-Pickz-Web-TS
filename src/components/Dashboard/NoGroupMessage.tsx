import React from 'react';
import styled from 'styled-components/macro';
import { Link } from 'react-router-dom';
import ActionButton from '../../Styles/Shared/ActionButton';
import { GroupSliderButtons } from '../../Styles/Groups';

const NoGroupMessage = () => {
  return (
    <Container>
      <TextContainer>
        You're ready to play, but you need to join a league first.
      </TextContainer>
      <GroupSliderButtons>
        <Link to='/groups/join'>
          <ActionButton>Join a League</ActionButton>
        </Link>
        <Link to='/groups/create'>
          <ActionButton>Create a League</ActionButton>
        </Link>
      </GroupSliderButtons>
    </Container>
  );
};

const Container = styled.div`
  display: grid;
  justify-content: center;
  grid-template-areas: '.' 'button';
  text-align: center;
  justify-items: center;
  grid-template-rows: 60vh 15vh;
  align-items: center;
  height: calc(100vh - 75px);
  width: 100vw;
`;
const TextContainer = styled.div`
  width: 60vw;
  font-size: 20px;
  line-height: 22px;
  letter-spacing: -0.04em;
`;

export default NoGroupMessage;
