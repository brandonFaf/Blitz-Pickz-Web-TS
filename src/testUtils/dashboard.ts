export const Dashboard0 = {
  rankings: [
    {
      points: 0,
      rank: 1
    }
  ]
};
export const Dashboard1 = {
  rankings: [
    {
      points: 8,
      rank: 2
    }
  ]
};
export const Dashboard7 = {
  rankings: [
    {
      points: 38,
      rank: 2
    }
  ]
};
export const Dashboard17 = {
  rankings: [
    {
      points: 140,
      rank: 1
    }
  ]
};
