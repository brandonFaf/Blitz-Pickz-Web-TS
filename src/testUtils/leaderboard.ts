import userData from './userData';
export const Leaderboard1 = {
  group_by_pk: {
    id: 1,
    display_name: 'test group 1'
  },
  rankings: [
    {
      points: 40,
      rank: 1,
      user: userData[0]
    }
  ]
};

export const Leaderboard2 = {
  group_by_pk: {
    id: 1,
    display_name: 'test group 2'
  },
  rankings: [
    {
      points: 40,
      rank: 1,
      user: userData[0]
    },
    {
      points: 38,
      rank: 2,
      user: userData[1]
    }
  ]
};
export const Leaderboard3 = {
  group_by_pk: {
    id: 1,
    display_name: 'Avengers'
  },
  rankings: [
    {
      points: 40,
      rank: 1,
      user: userData[0]
    },
    {
      points: 38,
      rank: 2,
      user: userData[1]
    },
    {
      points: 37,
      rank: 3,
      user: userData[2]
    },
    {
      points: 37,
      rank: 3,
      user: userData[3]
    },
    {
      points: 32,
      rank: 5,
      user: userData[4]
    },
    {
      points: 31,
      rank: 6,
      user: userData[5]
    },
    {
      points: 27,
      rank: 7,
      user: userData[6]
    }
  ]
};
