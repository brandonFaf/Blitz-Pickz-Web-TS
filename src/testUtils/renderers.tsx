//@ts-nocheck
import React from 'react';
import { MockedProvider } from '@apollo/client/testing';
import { UserContext } from '../contexts/UserContext';
import { GroupContext } from '../contexts/GroupContext';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
export const RenderApollo = ({
  children,
  mock = [],
  group = {},
  user = {}
}) => {
  const history = createMemoryHistory();
  return (
    <MockedProvider mocks={mock} addTypename={false}>
      <UserContext.Provider value={{ user }}>
        <GroupContext.Provider value={{ group }}>
          <Router history={history}>{children}</Router>
        </GroupContext.Provider>
      </UserContext.Provider>
    </MockedProvider>
  );
};
