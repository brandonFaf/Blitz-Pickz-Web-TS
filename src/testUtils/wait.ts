import { act } from 'react-dom/test-utils';

export default async () =>
  await act(async () => {
    await new Promise(resolve => setTimeout(resolve, 10));
  });
