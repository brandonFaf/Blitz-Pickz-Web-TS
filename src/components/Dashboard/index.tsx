import React, { useEffect } from 'react';
import getCurrentWeek from '../../helpers/currentWeek';
import Leaderboard from './Leaderboard';
import useGroup from '../../hooks/useGroup';
import NoGroupMessage from './NoGroupMessage';
import WeekStatus from '../../Styles/Dashboard/WeekStatus';
import styled from 'styled-components/macro';
import getOrdinal from '../../helpers/getOrdinal';
import MakePicks from './MakePicks';
import useUser from '../../hooks/useUser';
import { useDashbaordLazyQuery } from '../../types/graphql.types';

const Dashboard = () => {
  const { group } = useGroup();
  const { user } = useUser();

  const [
    loadDashboard,
    { data, loading, error, called }
  ] = useDashbaordLazyQuery();
  useEffect(() => {
    if (group) {
      loadDashboard({
        variables: { group_id: group.id, user_id: user.id }
      });
    }
  }, [group, loadDashboard, user.id]);
  if (loading) {
    return <div>Loading</div>;
  }
  if (error) {
    console.log(error);
    return <div>There was a problem loading this data</div>;
  }
  if (!group) {
    return <NoGroupMessage />;
  }
  if (!called) {
    return <div>Loading</div>;
  }
  return (
    <D>
      <DRel>
        <WeekStatus>
          <div className='week'>Week {getCurrentWeek()}</div>
          <div className='main'>
            You're in{' '}
            <strong>{getOrdinal(data?.rankings[0]?.rank || 1)}</strong> Place,
            <br />
            with{' '}
            <strong>
              {data?.rankings[0]?.points || 0}
              <sup>pts</sup>
            </strong>
          </div>
        </WeekStatus>
        <Leaderboard />
        <MakePicks />
      </DRel>
    </D>
  );
};

const D = styled.div`
  position: absolute;
`;
const DRel = styled.div`
  position: relative;
  width: 100vw;
  margin-left: auto;
  margin-right: auto;
  left: 0;
  right: 0;
  display: grid;
  grid-template-rows: 15vh 42vh;
`;
export default Dashboard;
