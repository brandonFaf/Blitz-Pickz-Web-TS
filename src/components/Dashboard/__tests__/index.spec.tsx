import React from 'react';
import { RenderApollo } from '../../../testUtils/renderers';
import Dashboard from '../index';
import { DashbaordDocument } from '../../../types/graphql.types';
import { Dashboard1 } from '../../../testUtils/dashboard';
import { render, cleanup } from '@testing-library/react';
import wait from '../../../testUtils/wait';
afterEach(cleanup);
jest.mock('../MakePicks.tsx');
jest.mock('../Leaderboard.tsx');
const renderLeaderboard = (mock, group, user) => {
  const props = { mock, group, user };
  return render(
    <RenderApollo {...props}>
      <Dashboard />
    </RenderApollo>
  );
};
describe('<Dashboard/>', () => {
  it('should render', async () => {
    const group = { id: 1 };
    const user = { id: '1' };
    const mock = [
      {
        request: {
          query: DashbaordDocument,
          variables: {
            group_id: group.id,
            user_id: user.id
          }
        },
        result: { data: Dashboard1 }
      }
    ];
    const { asFragment } = renderLeaderboard(mock, group, user);
    await wait();
    expect(asFragment()).toMatchSnapshot();
  });
  it('should render NoGroupMessage when there is no group', async () => {
    const group = null;
    const { getByText } = renderLeaderboard(undefined, group, undefined);
    await wait();
    expect(
      getByText("You're ready to play, but you need to join a league first.")
    );
  });
});
