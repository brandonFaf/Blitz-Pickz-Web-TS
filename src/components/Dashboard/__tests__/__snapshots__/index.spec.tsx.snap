// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard/> should render 1`] = `
<DocumentFragment>
  <div
    class="Dashboard__D-sc-13z2q2e-0 juNKjh"
  >
    <div
      class="Dashboard__DRel-sc-13z2q2e-1 cxmUfG"
    >
      <div
        class="WeekStatus-sc-1wmp7hy-0 jPRXJe"
      >
        <div
          class="week"
        >
          Week 1
        </div>
        <div
          class="main"
        >
          You're in 
          <strong>
            2nd
          </strong>
           Place,
          <br />
          with 
          <strong>
            8
            <sup>
              pts
            </sup>
          </strong>
        </div>
      </div>
      <div>
        Leaderboard
      </div>
      <div>
        Make Picks
      </div>
    </div>
  </div>
</DocumentFragment>
`;
